src/software/ai/passing/pass_generator.cpp

```diff
@@ -0,0 +1,179 @@
+    }
+
+    // Optimize the receiving positions for each robot and get the best pass
+    PassWithRating best_pass = optimizeReceivingPositions(world, receiving_positions_map);
+
+    // Visualize the sampled passes and the best pass
+    if (passing_config_.pass_gen_vis_config().visualize_sampled_passes())
+    {
+        std::vector<TbotsProto::DebugShapes::DebugShape> debug_shapes;
+        for (const auto& [robot_id, receiving_positions] : receiving_positions_map)
+        {
+            for (const Point& receiving_position : receiving_positions)
+            {
+                debug_shapes.push_back(*createDebugShape(
+                    Stadium(world.friendlyTeam().getRobotById(robot_id)->position(),
+                            receiving_position, 0.02),
+                    std::to_string(debug_shapes.size()) + "pg"));
+            }
+        }
+        std::stringstream stream;
+        stream << "BP:" << std::fixed << std::setprecision(3) << best_pass.rating;
+        debug_shapes.push_back(
+            *createDebugShape(Circle(best_pass.pass.receiverPoint(), 0.05),
+                              std::to_string(debug_shapes.size()) + "pg", stream.str()));
+        LOG(VISUALIZE) << *createDebugShapes(debug_shapes);
+    }
+
+    // Generate sample passes across the field for cost visualization
+    if (passing_config_.cost_vis_config().generate_sample_passes())
+    {
+        samplePassesForVisualization(world, passing_config_, best_pass.pass);
+    }
+
+    return best_pass;
+}
+
+std::map<RobotId, std::vector<Point>> PassGenerator::sampleReceivingPositionsPerRobot(
+    const World& world, const std::vector<RobotId>& robots_to_ignore)
+{
+    std::map<RobotId, std::vector<Point>> receiving_positions_map;
+
+    const double min_sampling_std_dev =
+        passing_config_.pass_gen_min_rand_sample_std_dev_meters();
+    const double sampling_std_dev_vel_multiplier =
+        passing_config_.pass_gen_rand_sample_std_dev_robot_vel_multiplier();
+    const double sampling_center_vel_multiplier =
+        passing_config_.pass_gen_rand_sample_center_robot_vel_multiplier();
+
+    for (const Robot& robot : world.friendlyTeam().getAllRobots())
+    {
+        // Ignore robots in the ignore list
+        if (std::find(robots_to_ignore.begin(), robots_to_ignore.end(), robot.id()) !=
+            robots_to_ignore.end())
+        {
+            continue;
+        }
+
+        // Add the robot's current position to the list of sampled passes
+        auto robot_position = robot.position();
+        receiving_positions_map.insert({robot.id(), {robot_position}});
+
+        // Add the best pass from the previous iteration to the list of sampled passes
+        if (previous_best_receiving_positions_.find(robot.id()) !=
+            previous_best_receiving_positions_.end())
+        {
+            receiving_positions_map[robot.id()].push_back(
+                previous_best_receiving_positions_[robot.id()]);
+        }
```

Comment:
Why do we need this? Aren't we just calling this function once at line 19?

Reply:
Also, isn't it already appended in line 167?

Reply:
Every time `PassGenerator::getBestPass` is called, we randomly sample a set of new receiving positions and optimize them. One iteration we might randomly sample a really good receiving positions, while the next iteration we might get unlucky with the RNG and not sample any receiving positions close to the really good receiving position. The idea with `previous_best_receiving_positions_` is to cache the best _sampled_ receiving positions from one iteration to the next so even if we get unlucky RNG, we still have the really good receiving position from last iteration. This also helps stabilize the receiving positions between consecutive iteration, without the receiving position jumping around with every call. 

Line 167 is the part where we store the best receiving position (if it has a decent score i.e. >=0.1) for the next iteration. Line 91 is where we use the best receiving positions from the previous iteration.

Hope this clarifies the logic here. Let me know if it's still confusing.